// @zondax/zemu/dist/src/constants.js
"use strict";
exports.__esModule = true;
exports.WINDOW_X = exports.WINDOW_S = exports.KEYS = exports.KILL_TIMEOUT = exports.BASE_NAME = exports.DEFAULT_TRANSPORT_PORT = exports.DEFAULT_VNC_PORT = exports.DEFAULT_HOST = exports.DEFAULT_KEY_DELAY = exports.DEFAULT_START_DELAY = exports.DEFAULT_MODEL = exports.TIMEOUT = exports.DEFAULT_EMU_IMG = void 0;
exports.DEFAULT_EMU_IMG = 'zondax/builder-zemu@sha256:21121adcc236ef230e3e2e64940727abc1f1c9539cbdcffddb6e0fecc5b977d9';
exports.TIMEOUT = 1000;
exports.DEFAULT_MODEL = 'nanos';
exports.DEFAULT_START_DELAY = 3500;
exports.DEFAULT_KEY_DELAY = 350;
exports.DEFAULT_HOST = '127.0.0.1';
exports.DEFAULT_VNC_PORT = 8001;
exports.DEFAULT_TRANSPORT_PORT = 9998;
exports.BASE_NAME = 'zemu-656d75-';
exports.KILL_TIMEOUT = 10000;
exports.KEYS = {
    NOT_PRESSED: 0,
    PRESSED: 1,
    LEFT: 0xff51,
    RIGHT: 0xff53
};
exports.WINDOW_S = {
    x: 0,
    y: 0,
    width: 128,
    height: 32
};
exports.WINDOW_X = {
    x: 0,
    y: 0,
    width: 128,
    height: 64
};
